controlplane/observability.go

```diff
@@ -60,67 +60,56 @@ func (o *ObservabilityService) SendMetric(ctx context.Context, req *connect.Requ
 	return connect.NewResponse(&ftlv1.SendMetricResponse{}), nil
 }
 
-func (o *ObservabilityService) insertSumMetrics(ctx context.Context, runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Sum) error {
-	for _, dataPoint := range metric.Sum.DataPoints {
-		err := o.dal.InsertMetricEntry(ctx, dal.Metric{
-			RunnerKey:  runnerKey,
-			StartTime:  time.Unix(0, int64(dataPoint.StartTimeUnixNano)),
-			EndTime:    time.Unix(0, int64(dataPoint.TimeUnixNano)),
-			SourceVerb: getVerbRef(dataPoint.Attributes, true),
-			DestVerb:   getVerbRef(dataPoint.Attributes, false),
-			Name:       name,
-			DataPoint:  dal.MetricCounter{Value: dataPoint.GetAsInt()},
+func (o *ObservabilityService) extractSumMetrics(runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Sum) []dal.Metric {
+	return slices.Map(metric.Sum.DataPoints, func(dataPoint *metricsv1.NumberDataPoint) dal.Metric {
+		return newMetric(runnerKey, name, dataPoint, dal.MetricCounter{
+			Value: dataPoint.GetAsInt(),
 		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
-	}
-	return nil
+	})
 }
 
-func (o *ObservabilityService) insertHistogramMetrics(ctx context.Context, runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Histogram) error {
-	for _, dataPoint := range metric.Histogram.DataPoints {
-		err := o.dal.InsertMetricEntry(ctx, dal.Metric{
-			RunnerKey:  runnerKey,
-			StartTime:  time.Unix(0, int64(dataPoint.StartTimeUnixNano)),
-			EndTime:    time.Unix(0, int64(dataPoint.TimeUnixNano)),
-			SourceVerb: getVerbRef(dataPoint.Attributes, true),
-			DestVerb:   getVerbRef(dataPoint.Attributes, false),
-			Name:       name,
-			DataPoint: dal.MetricHistogram{
-				Count:  int64(dataPoint.Count),
-				Sum:    int64(*dataPoint.Sum),
-				Bucket: dataPoint.BucketCounts,
-			},
+func (o *ObservabilityService) extractHistogramMetrics(runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Histogram) []dal.Metric {
+	return slices.Map(metric.Histogram.DataPoints, func(dataPoint *metricsv1.HistogramDataPoint) dal.Metric {
+		return newMetric(runnerKey, name, dataPoint, dal.MetricHistogram{
+			Count:  int64(dataPoint.Count),
+			Sum:    int64(*dataPoint.Sum),
+			Bucket: dataPoint.BucketCounts,
 		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
-	}
-	return nil
+	})
 }
 
-func getVerbRef(attributes []*commonv1.KeyValue, isSource bool) schema.VerbRef {
-	var name string
-	var module string
-
+func fillMetricFromAttributes(metric *dal.Metric, attributes []*commonv1.KeyValue) {
```

Comment:
Similar to our optimisation today, this avoids iterating across the attributes once for each value we're extracting, instead iterating once.